From 6696121faeefb28f76c69ef97c3b4f20f5d75d68 Mon Sep 17 00:00:00 2001
From: Tomi Ollila <tomi.ollila@iki.fi>
Date: Wed, 12 Aug 2020 23:40:13 +0300
Subject: [DRAFT PATCH V2] emacs: show local date next to Date: in case value differs

When adding Date: header of a message to notmuch-show buffer, compare the
date string with local representation of it and if these differ, output
Date: {original-date-string}  ({local-date-representation})

This is useful when mail system provides Date: strings with
different timezone information than the sender is located at.

---

V2 of id:1427132722-20346-1-git-send-email-tomi.ollila@iki.fi
sent Mon, 23 Mar 2015 19:45:22 +0200.

Resent due to conflict during rebase Aug 2020.

And still an "early" draft =D, see discussion at

https://nmbug.notmuchmail.org/nmweb/search/id%3A1427132722-20346-1-git-send-email-tomi.ollila%40iki.fi

 emacs/notmuch-show.el | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/emacs/notmuch-show.el b/emacs/notmuch-show.el
index b0f2d28b..6498c041 100644
--- a/emacs/notmuch-show.el
+++ b/emacs/notmuch-show.el
@@ -482,14 +482,27 @@ (defun notmuch-show-insert-header (header header-value)
   "Insert a single header."
   (insert header ": " (notmuch-sanitize header-value) "\n"))
 
-(defun notmuch-show-insert-headers (headers)
+(defun notmuch--make-date (timestamp)
+  (if (> timestamp 2147483647)
+      (message-make-date (seconds-to-time timestamp))
+    (message-make-date (encode-time timestamp 0 0 1 1 1970 t))))
+
+(defun notmuch-show-insert-headers (headers &optional timestamp)
   "Insert the headers of the current message."
-  (let ((start (point)))
+  (let ((start (point))
+	date-local)
     (mapc (lambda (header)
 	    (let* ((header-symbol (intern (concat ":" header)))
 		   (header-value (plist-get headers header-symbol)))
 	      (when (and header-value
 			 (not (string-equal "" header-value)))
+		(if (and timestamp
+			 (string-equal header "Date")
+			 (not (string-equal
+			       (setq date-local (notmuch--make-date timestamp))
+			       header-value)))
+		    (setq header-value
+			  (format "%s  (%s)" header-value date-local)))
 		(notmuch-show-insert-header header header-value))))
 	  notmuch-message-headers)
     (save-excursion
@@ -1056,7 +1069,7 @@ (defun notmuch-show-insert-msg (msg depth)
     ;; Set `headers-start' to point after the 'Subject:' header to be
     ;; compatible with the existing implementation. This just sets it
     ;; to after the first header.
-    (notmuch-show-insert-headers headers)
+    (notmuch-show-insert-headers headers (plist-get msg :timestamp))
     (save-excursion
       (goto-char content-start)
       ;; If the subject of this message is the same as that of the
-- 
2.25.1

